Various changes to card editor.

thomasloven opened the pull request:
Various tweaks and fixes to the card editor.

thomasloven committed:
Various changes to card editor.

thomasloven commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+  public set yaml(_yaml: string) {
+    this._yaml = _yaml;
+    this._updateConfigElement();
+    // TODO: Defer this to next redraw? How?
```
RFC!

balloob replied:
To defer something after a draw:
```ts
await this.updateComplete
```

balloob replied:
But you cannot do that in a setter since that's not async. However, since `this.updateComplete` is a promise, you can always `.then` it.

balloob replied:
Alternative is that you could check inside `updated` if `_yaml` has been updated, and if so, call this.

thomasloven replied:
Ok. This just doesn't want to work unless I use this 1ms delay... Tried `updateComplete.then`, tried `updated()`, it all breaks...

thomasloven commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+    }
+  }
+
+  private _handleUIConfigChanged(ev) {
```
What type should `ev` have?

balloob replied:
define a new type.

Example from `polymer-types.ts`:

```ts
export interface PolymerChangedEvent<T> extends Event {
  detail: {
    value: T;
    path?: string;
    queueProperty: boolean;
  };
}
```

thomasloven commented on src/panels/lovelace/editor/card-editor/hui-card-picker.ts:
```diff
@@ -97,7 +107,7 @@ export class HuiCardPicker extends LitElement {
       config = { ...config, ...cardConfig };
     }
 
-    this.cardPicked!(config);
```
I read somewhere that callbacks are discouraged. Correct?

balloob replied:
Correct.

thomasloven commented:
![Skärminspelning 2019-06-12 kl  22 25 06 mov](https://user-images.githubusercontent.com/1299821/59384547-b7b64d00-8d62-11e9-92aa-b2fc3b7da896.gif)

thomasloven committed:
Avoid crashing on bad yaml when creating a new card

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+    this._updateConfigElement();
+    // TODO: Defer this to next redraw? How?
+    setTimeout(() => this._yamlEditor.codemirror.refresh(), 1);
+    fireEvent(this, "config-changed", { config: this.config! });
```
If we want to follow how DOM elements fire change events, we should not trigger when the value is set from the outside, but only when it is changed from within the element.

balloob replied:
Otherwise we can get into loops.

thomasloven replied:
I finally realized the obvious way to fix this.
Yaml is still the main source of truth, but each time it is changed, it's also parsed into a configuration. If invalid, the old configuration is still stored, but an error is sent to the host element.
When updating the configuration (which is now known as `value`) from outside, it's checked against this stored value, and if it's equal, nothing is changed. Otherwise, it replaces the yaml.

Returning and being fed back bad values avoided. Loops gone. I hope.

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+
+  protected render(): TemplateResult {
+    return html`
+      <div style="width: 100%;">
```
we should not use inline CSS

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+  protected render(): TemplateResult {
+    return html`
+      <div style="width: 100%;">
+        <div class="gui-editor" ?hidden="${this._GUImode === false}">
```
Why use `hidden` and not use a turnary statement like below for loading?

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+          ${this._loading
+            ? html`
+                <paper-spinner
+                  ?active="${this._loading}"
```
No need to make `active` conditional since you only render this when loading.

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+        </div>
+        <div class="yaml-editor" ?hidden="${this._GUImode === true}">
+          <hui-yaml-editor
+            .hass="${this.hass}"
```
No need to use quotes around the `${…}` values

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+            .value="${this.yaml}"
+            @yaml-changed="${this._handleYAMLChanged}"
+          ></hui-yaml-editor>
+          <div class="error">
```
Should these be conditionally rendered?

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+    this.yaml = yaml.safeDump(config);
+  }
+
+  public get error(): boolean {
```
It's weird that this is a boolean but not stored as one. I think that we should rename it to `hasError`

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+        </div>
+        <div class="buttons">
+          <mwc-button
+            @click="${() => (this._GUImode = !this._GUImode)}"
```
We should not define inline functions. Instead make it a class method. 

Inline functions will be re-created each time render is called and will always cause an update, because the newly created object has a different identity than the old function.

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+            ?disabled="${this._warning || this._error}"
+            ?unelevated="${this._GUImode === false}"
+          >
+            <ha-icon .icon="${"mdi:code-braces"}"></ha-icon>
```
You should be able to do `icon="mdi:code-braces"`

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+  static get styles(): CSSResult {
+    return css`
+      :host {
+        margin: 0 10px;
```
We should never add margin to our elements. Margin has to be applied by whoever element embeds this. We don't know how this will be used, and so can't decide that a margin is always necessary. If a margin is necessary for rendering, we should accomplish that with a padding.

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+        text-align: right;
+
+        padding: 8px 0px;
+        border-top: 1px solid #e8e8e8;
```
CSS vars?

thomasloven replied:
Forgotten debug line

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+        this._yamlEditor.codemirror.refresh();
+        this._yamlEditor.codemirror.focus();
+      }
+      fireEvent(this as HTMLElement, "iron-resize");
```
Why this casting to HTMLElement?

thomasloven replied:
🤷‍♂ Old code did it

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+  private _handleUIConfigChanged(ev) {
+    ev.stopPropagation();
+    const config = ev.detail.config;
+    if (JSON.stringify(config) !== JSON.stringify(this.config)) {
```
why would we do this expensive computation ? We should just apply it.

thomasloven replied:
My thinking was this could be used to check for changes that needed saving. But that's better handled somewhere else... and also a thing that I'd forgotten about entirely...

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+    `;
+  }
+
+  static get styles(): CSSResult {
```
Could you move this to be the last method in the class? We have done this for other classes and it has helped readability (because we don't style as much as that we tweak callbacks)

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+  }
+
+  protected async firstUpdated() {
+    this.config = this.value;
```
Is this necessary? Shouldn't we define a getter that is `this.config || this.value` ?

thomasloven replied:
Turns out the `value` property can be removed and `config` used in the interface.
That didn't work before because the editor was embedded in an element that kept updating and overwriting the changes.

thomasloven replied:
Nope. I was wrong. This is needed.

`value` needs to be transfered to `config` only once. Otherwise it will be overwritten by the embedding element when an error is returned (from `get config()`).

`this.config || this.value` won't work, because `this.config` just points to `this.yaml` and doing this for that would get weird in the off chance that a user opens the yaml editor and clears it.

balloob replied:
This does add the side effect that it is no longer possible to externally update the value of this element.

balloob replied:
I think that we should move this to `updated` and do it each time `value` has changed

balloob commented on src/panels/lovelace/editor/card-editor/hui-card-editor.ts:
```diff
@@ -0,0 +1,264 @@
+      this._loading = false;
+      fireEvent(this as HTMLElement, "iron-resize");
+    }
+    return;
```
not needed

balloob commented on src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:
```diff
@@ -33,72 +41,211 @@ export class HuiDialogEditCard extends LitElement {
 
-  private _cardPicked(cardConf: LovelaceCardConfig): void {
-    this._cardConfig = cardConf;
+  static get styles(): CSSResult[] {
```
Use `CSSResultArray`, the result is actually flattened so the `CSSResultArray` type allows that.

balloob commented on src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:
```diff
@@ -33,72 +41,211 @@ export class HuiDialogEditCard extends LitElement {
+          height: 14px;
+          margin-right: 20px;
+        }
+        paper-spinner {
```
We shouldn't do this, we should prefer to use ternaries in our templates.

thomasloven committed:
Address review comments

thomasloven commented:
"Various changes" comprises:

- Separation of concerns. `hui-card-editor` now gets a card configuration, displays a GUI or yaml editor, and returns a card configuration. `hui-card-preview` gets a card configuration and displays a card. They shouldn't need to know about each other or where the configuration is stored/saved to. This also makes `hui-card-editor` recurseable, which might help when trying to solve the stack editor problem.
- Make the editor use a single source of truth for the configuration. Before it was converted between yaml and json when changing between gui and yaml mode. Now the configuration is always stored as yaml and converted to and from json when needed (I chose to do it this direction to guarantee yaml formating was preserved).
- Remove the pick-card dialog and make it a function of the edit-card dialog instead.
- Trying to clean up the code and DOM (highly subjective) from the MVP-state it was in.

Left to do:

- Display a warning when trying to close the dialog with unsaved changes.

thomasloven committed:
Revert interface change

balloob commented:
I still see 2 of my comments unaddressed  (either by a code change or a comment saying why not)

thomasloven commented:
I held off a bit on them because I didn't immediately realize how to solve it. May need a little bit of restructuring.
Vacation just started, so I'll take another look at it very soon.

balloob commented:
Alright! Was just checking in .

thomasloven committed:
Avoid config loops. Nicer error behavior.

thomasloven commented:
Rebased. Fixed some types. Restructured handling and passing of configuration back and forth.
Improved error behavior (when the yaml can't be parsed).
![Skärminspelning 2019-08-08 kl  19 55 33 mov](https://user-images.githubusercontent.com/1299821/62727140-17566f00-ba19-11e9-9548-da560ee2ae7f.gif)

balloob commented:
Still WIP ?

thomasloven commented:
No. I'm pretty happy with it now.
Only thing is that the symbol for switching between GUI and YAML may not be quite as intuitive as I'd hoped, and it may not be quite clear why it sometimes is unclickable.

balloob commented:
🎉
